File: linked_list/c/ex_4.c
/**
 * Partition: Write code to partition a linked list around a value x, such that
 * all nodes less than x come before all nodes greater than or equal to x. If
 * x is contained within the list, the values of x only need to be after the
 * elements less than x. The partition element x can appear anywhere in the
 * "right partition"; it does not need to appear between the left and right
 * paritions.
 */

#include <stdio.h>
#include "linked_list.h"

list *partition(list *list_p, int partition);

int main() {
    list *test_list = create_list();
    insert_multiple(test_list, 15, insert_unsorted, 3, 6, 5, 7, 8, 2, 5, 0, 1, 3, 4, 7, 4, 2, 9);
    partition(test_list, 5);
    print_list(test_list);
}

list *partition(list *list_p, int partition) {
    if (list_p->size < 2) return list_p;

    node *less, *more;
    less = more = list_p->head;

    while(1) {
        while (less != NULL && less->data >= partition) {
            less = less->next;
            printf("less->data = %i\n", less->data);
        }

        while (more != NULL && more->data <= partition) {
            more = more->next;
            printf("more->data = %i \n", more->data);
        }

        if (more == NULL || less == NULL) break;

        printf("Swapping less = %i and more = %i\n", less->data, more->data);
        int temp = less->data;
        less->data = more->data;
        more->data = temp;
    }

    return list_p;
}

File: go/linked_lists/ctci/ex4.go
package ctci

// 1.4 Partition: Write code to partition a linked list around a value x, such
// that all the nodes less than x come before all nodes greater than or equal
// to x. If x is contained within the list, the values of x only need to be
// after the elements less than x. The partition element x can appear anywhere
// in the right partition; it does not need to appear between th left and right
// partitions.
//
// EXAMPLE
// Input: 3 -> 5 -> 8 -> 5 -> 10 -> 2 -> 1 [partition = 5]
// Output: 3 -> 1 -> 2 -> 10 -> 5 -> 5 -> 8
//
// Assumptions:
//	* Singly linked list
//  * Confirm whether partition should be stable or unstable
//	* List does not contain cycles

// PartitionStable constructs two linked lists, one containing elements less
// than val, and the other containing all other elements, then joins the lists
// together, preserving the relative ordering of of elements in each partition.
//
// Time complexity: O(n)
// Space complexity: O(1)
func PartitionStable(s *SNode, val int) *SNode {
	var firstHalfHead, firstHalfTail, secondHalfHead, secondHalfTail *SNode
	for cur := s; cur != nil; {
		// Setting each cur.Next to nil prevents needings to check each half
		// for nil and setting the appropriate tail to nil when the loop exits.
		next := cur.Next
		cur.Next = nil
		if cur.Data < val {
			appendToTail(firstHalfHead, firstHalfTail, cur)
		} else {
			appendToTail(secondHalfHead, secondHalfTail, cur)
		}
		cur = next
	}

	// Ensure first half is not nil before attempting to join lists together.
	if firstHalfHead == nil {
		return secondHalfHead
	}

	firstHalfTail.Next = secondHalfHead
	return firstHalfHead
}

func appendToTail(head, tail, n *SNode) {
	if head == nil {
		head = n
		tail = n
	} else {
		tail.Next = n
		tail = n
	}
}

// PartitionUnstable appends to the head of a new list if the node's value is
// less than the partition value, and to the tail if it is equal to or greater.
//
// Time complexity: O(n)
// Space complexity: O(1)
func PartitionUnstable(n *SNode, val int) *SNode {
	head := n
	tail := head
	for cur := head.Next; cur != nil; {
		next := cur.Next
		if cur.Data < val {
			cur.Next = head
			head = cur
		} else {
			tail.Next = cur
			tail = cur
		}

		cur = next
	}
	tail.Next = nil

	return head
}
